### DISCUSSION TITLE
Bad CRC when zipping

### DISCUSSION THREAD

**Discussion by codydietrich** [101arrowz/fflate#40] [Q&A] [answered]
Hi, I just found this library as I was looking for a faster alternative to JSZip. I've implemented fflate into my code, and using the same test set of images to zip, the resulting zip is not valid. Running `unzip` reports `bad CRC` and each file is a bit corrupted. I'm using the following:
```javascript
const zip = new fflate.Zip();
const zipReadableStream = fflToRS(zip);
for (const file of resp.files) {
    const ext = file.filename.slice(file.filename.lastIndexOf('.') + 1).toLowerCase();
    const zipFile = new fflate.ZipDeflate(file.filename, {
      level: ALREADY_COMPRESSED.indexOf(ext) == -1 ? 6 : 0
    });
    zip.add(zipFile);
    await downloadAndDecrypt(zipFile, file.url, file.key);
}
zip.end();
const zipResponse = new Response(zipReadableStream);
saveAs(await zipResponse.blob(), "test.zip");
```

`downloadAndDecrypt` uses `fetch` to download an image, read it's readable stream into 64k chunks, decrypt, and push the resulting chunk to `zipFile` - `zipFile.push(chunk)` and the last chunk is pushed as `zipFile.push(chunk, true)`.

There's no issue with the download and decrypt part, as it works fine (albeit a bit slower) with `JSZip`.

**Comment by 101arrowz** [comment, marked as answer]
I might be able to fix the bug, but first you need to check a few things:
- Are you using the latest version of `fflate` (v0.6.3)?
- Does the CRC value appear to be unset, i.e. is the value in the ZIP file 00000000?
- Have you tried `ZipPassThrough` over `ZipDeflate` (to see if it's the compression or something else).
- Could you show the source code for `downloadAndDecrypt`?

By the way, if you want to improve performance, use parallelized downloading. Don't `await` the `downloadAndDecrypt` function so that multiple files can be downloaded in parallel; use `AsyncZipDeflate` over `ZipDeflate` to compress in parallel as well.

FWIW, I was able to get URL-downloading, streaming ZIP compression working with the code in the following spoiler and the HTML from the buildless example:

<details>
<summary>Go <a href="https://cors-anywhere.herokuapp.com/corsdemo">here</a> before running this code</summary>

```js
import * as fflate from 'https://cdn.skypack.dev/fflate@0.6.3?min';
const zipDownloadList = document.getElementById('zip-download-list');

const handleFiles = (urls) => {
  const zip = new fflate.Zip();
  const zipReadableStream = fflToRS(zip);
  for (const url of urls) addFileToZip(url, zip);
  zip.end();
  const zipResponse = new Response(zipReadableStream);
  addDownloadButton(zipResponse, `fflate-zip-${urls[0]}.zip`);
}

const fflToRS = fflateStream =>
  new ReadableStream({
    start(controller) {
      fflateStream.ondata = (err, data, final) => {
        if (err) controller.error(err);
        else {
          controller.enqueue(data);
          if (final) controller.close()
        }
      }
    },
    cancel() {
      fflateStream.terminate();
    }
  });

const incompressibleTypes = new Set([
  'zip', 'gz', 'png', 'jpg', 'jpeg', 'pdf', 'doc', 'docx',
  'ppt', 'pptx', 'xls', 'xlsx', 'heic', 'heif', '7z', 'bz2',
  'rar', 'gif', 'webp', 'webm', 'mp4', 'mov', 'mp3', 'aifc'
]);

const addFileToZip = async (url, zip) => {
  const filename = url.slice(url.lastIndexOf('/') + 1);
  const ext = filename.slice(filename.lastIndexOf('.') + 1);
  const zippedFileStream = incompressibleTypes.has(ext)
    ? new fflate.ZipPassThrough(filename)
    : new fflate.AsyncZipDeflate(filename, { level: 9 });
  zip.add(zippedFileStream);
  const req = await fetch(`https://cors-anywhere.herokuapp.com/${url}`);
  const fileReader = req.body.getReader();
  while (true) {
    const { done, value } = await fileReader.read();
    if (done) {
      zippedFileStream.push(new Uint8Array(0), true);
      break;
    }
    zippedFileStream.push(value);
  }
}

const addDownloadButton = async (response, filename) => {
  const blob = await response.blob();
  const downloadURL = URL.createObjectURL(blob);
  const downloadLink = document.createElement('a');
  downloadLink.href = downloadURL;
  downloadLink.download = filename;
  const downloadButton = document.createElement('button');
  downloadButton.textContent = `Download ${filename}`;
  downloadButton.addEventListener('click', event => {
    downloadLink.click();
  });
  const downloadElement = document.createElement('li');
  downloadElement.appendChild(downloadButton);
  zipDownloadList.appendChild(downloadElement);
}

// Use whatever URLs you want
handleFiles([
  'https://www.gutenberg.org/files/2701/old/moby10b.txt',
  'https://hlevkin.com/hlevkin/TestImages/new/Rainier.bmp'
]);
```
</details>

**Reply by codydietrich** [reply]
Hi, thanks for quick reply.
- Yes, latest version
- No, CRC is reported as incorrect - `bad CRC 6d77087e  (should be f60372cc)` for example
- This seems to solve the issue. When using your suggested method of `ZipPassThrough` and `AsyncZipDeflate` the issue is fixed. Maybe there is a bug somewhere within `ZipDeflate`?

Thanks for the help, it is working well now, along with the parallelized downloading.

**Reply by 101arrowz** [reply]
That's quite interesting, that means there's some bug with CRC calculation for the `ZipDeflate` but somehow not for `AsyncZipDeflate`, even though they differ by very little. I'd appreciate if you could send me files that can reproduce this bug so that I can fix it for future users.
